### PR TITLE
Add diag for coap latency

### PR DESCRIPTION
**Pull request by keeramis**
## NOTE: this PR targets ch47377/CoAP_msg_retr_counter_deviceos

<details>
  <summary><i>submission notes</i></summary>

```
**Important:** Please sanitize/remove any confidential info like usernames, passwords, org names, product names/ids, access tokens, client ids/secrets, or anything else you don't wish to share.

Please Read and Sign the Contributor License Agreement ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md)).

You may also delete this submission notes header if you'd like. Thank you for contributing!
```
</details>

### Problem

Currently, there is no diagnostic to show the coap-latency of the messages going outbound from the device to the cloud. 

### Solution

Add diagnostic to calculate the coap latency of the last message sent from device to cloud (outbound messages). Measured for each message using the `id`

### Example App

```c
#include "application.h"

SerialLogHandler logHandler(LOG_LEVEL_ALL);

auto diag_id_coap_latency = DIAG_ID_CLOUD_COAP_LATENCY;
AbstractIntegerDiagnosticData::IntType val_coap_latency;

void setup() {
}

void loop() {   
    delay(1000);
    
    char buf[5] = {0};
    sprintf(buf,"%d",i);
    Particle.publish("dummy", buf, PRIVATE);
    
    AbstractIntegerDiagnosticData::get(diag_id_coap_latency, val_coap_latency);
    Serial.printlnf("[TEST] COAP LATENCY FOR THE LAST MSG : %d", val_coap_latency);
}
```

### References
- Corresponding change's PR on the device-service side: https://github.com/particle-iot/device-service/pull/308
- https://app.clubhouse.io/particle/story/47381/measure-the-time-between-the-original-message-transmission-and-the-acknowledgement
- https://app.clubhouse.io/particle/story/47377/measure-number-of-coap-message-retransmissions

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

**Comment by keeramis** [comment]
- Suggestions welcome to how to add unit tests to this particular metric

- This is not branched off of develop, and is instead branched off of https://github.com/particle-iot/device-os/tree/ch47377/CoAP_msg_retr_counter_deviceos, to be able to use `SimpleUnsignedIntegerDiagnosticData` which is its `unsigned` variant,  which is not available in `develop` yet

**Comment by keeramis** [comment]
Results: 

From the app above, a message is sent to cloud every 1 sec. And coap latency is measured each time, and recorded. Following is data from 4255 values (~70min of testing)

image

File: communication/src/coap_channel.cpp [hunk 1]
```diff
@@ -124,15 +131,17 @@ ProtocolError CoAPMessageStore::receive(Message& msg, Channel& channel, system_t
...
+		CoAPMessage* coap_msg = from_id(id);
+		if (coap_msg)
+		{
+			g_coapLatencyMSec = time - coap_msg->get_send_time();
+			if (msgtype==CoAPType::RESET) {
+				coap_msg->notify_delivered_nak();
+				// a RESET indicates that the session is invalid.
+				// Currently the device never sends a RESET, but if it were to do that
+				// then we should track which direction we are sending
+				channel.command(Channel::DISCARD_SESSION, nullptr);
```
Comment (sergeuz) [inline communication/src/coap_channel.cpp]: This changes how `RST` messages are handled. Previously, the `DISCARD_SESSION` command would be called even if the original CoAP message was not found.

Comment (keeramis) [reply communication/src/coap_channel.cpp]: @sergeuz great catch. thank u for checking. 
Changed! There are multiple ways to refactor this little logic- I just separated both the if statements. Please check.

File: communication/src/communication_diagnostic.h [hunk 2]
```diff
@@ -3,4 +3,5 @@
 extern particle::SimpleUnsignedIntegerDiagnosticData g_rateLimitedEventsCounter;
 extern particle::SimpleUnsignedIntegerDiagnosticData g_unacknowledgedMessageCounter;
 extern particle::SimpleUnsignedIntegerDiagnosticData g_trasmittedMessageCounter;
-extern particle::SimpleUnsignedIntegerDiagnosticData g_retransmittedMessageCounter;
\ No newline at end of file
+extern particle::SimpleUnsignedIntegerDiagnosticData g_retransmittedMessageCounter;
+extern particle::SimpleUnsignedIntegerDiagnosticData g_coapRoundTripMSec;
```
Comment (eberseth) [inline communication/src/communication_diagnostic.h]: Add newline

File: communication/src/coap_channel.cpp [hunk 3]
```diff
@@ -124,10 +131,13 @@ ProtocolError CoAPMessageStore::receive(Message& msg, Channel& channel, system_t
 	if (msgtype==CoAPType::ACK || msgtype==CoAPType::RESET)
 	{
 		message_id_t id = msg.get_id();
+		CoAPMessage* coap_msg = from_id(id);
+		if (coap_msg) {
+			g_coapRoundTripMSec = time - coap_msg->get_send_time();
```
Comment (eberseth) [inline communication/src/coap_channel.cpp]: Will this be safe for rollovers?

File: communication/src/communication_diagnostic.cpp [hunk 4]
```diff
@@ -3,4 +3,5 @@
 particle::SimpleUnsignedIntegerDiagnosticData g_rateLimitedEventsCounter(DIAG_ID_CLOUD_RATE_LIMITED_EVENTS, DIAG_NAME_CLOUD_RATE_LIMITED_EVENTS);
 particle::SimpleUnsignedIntegerDiagnosticData g_unacknowledgedMessageCounter(DIAG_ID_CLOUD_UNACKNOWLEDGED_MESSAGES, DIAG_NAME_CLOUD_UNACKNOWLEDGED_MESSAGES);
 particle::SimpleUnsignedIntegerDiagnosticData g_trasmittedMessageCounter(DIAG_ID_CLOUD_TRANSMITTED_MESSAGES, DIAG_NAME_CLOUD_TRANSMITTED_MESSAGES);
-particle::SimpleUnsignedIntegerDiagnosticData g_retransmittedMessageCounter(DIAG_ID_CLOUD_RETRANSMITTED_MESSAGES, DIAG_NAME_CLOUD_RETRANSMITTED_MESSAGES);
\ No newline at end of file
+particle::SimpleUnsignedIntegerDiagnosticData g_retransmittedMessageCounter(DIAG_ID_CLOUD_RETRANSMITTED_MESSAGES, DIAG_NAME_CLOUD_RETRANSMITTED_MESSAGES);
+particle::SimpleUnsignedIntegerDiagnosticData g_coapRoundTripMSec(DIAG_ID_CLOUD_COAP_ROUND_TRIP, DIAG_NAME_CLOUD_COAP_ROUND_TRIP);
```
Comment (eberseth) [inline communication/src/communication_diagnostic.cpp]: Add newline

File: communication/src/coap_channel.cpp (hunk 3 above)
Comment (keeramis) [reply communication/src/coap_channel.cpp]: Correct. This is associated with `SimpleUnsignedIntegerDiagnosticData` which now handles rollovers... 
This is the PR where rollovers are added - https://github.com/particle-iot/device-os/pull/2043

**Pull request merged**